### PR TITLE
Use babashka script

### PR DESCRIPTION
**Pull request by porkostomus**
I figured out how to override the methods in `clojure.test` to output the results of every test.

This avoids having to use another testing framework which was overkill for our purpose. I was only leaning on it because I didn't know what I was doing.

This also avoids needing to use an image with the JVM at all, since babashka ships as a small self-contained binary that runs on Alpine. It uses an interpreted version of Clojure which for the purposes of Exercism will be totally fine.

Best of all, we won't be paying the JVM startup time every time it is run, which may not have been so practical for giving immediate feedback.

The script, `test-runner.clj`, takes the 3 command-line args and outputs to `<out-dir>/results.json` an object with the following keys:

```bash
$ test-runner.clj two-fer in-dir/ out-dir/
```

```json
{
  "fails" : [ ],
  "errors" : [ ],
  "passes" : [ "two-fer-test", "name-alice-test", "name-bob-test" ],
  "status" : "pass",
  "version" : 2
}
```

The output obviously does not yet match the spec, but that will most likely be another PR.

### CODE REVIEW

**Review by ErikSchierboom — COMMENTED** [review]
> This also avoids needing to use an image with the JVM at all, since babashka ships as a small self-contained binary that runs on Alpine. It uses an interpreted version of Clojure which for the purposes of Exercism will be totally fine.
> 
> Best of all, we won't be paying the JVM startup time every time it is run, which may not have been so practical for giving immediate feedback.

Excellent!

> The output obviously does not yet match the spec, but that will most likely be another PR.

A different PR is good, but we can't really merge this PR right now as having the output not be according to the spec would break the current, working test runner.

**Comment by porkostomus** [comment]
All this PR does is add files. The current test runner still works, I just checked to make sure.

**Comment by porkostomus** [comment]
I think I've got it much closer! I created an alternate run script called `run-bb.sh` which instead calls the babashka script and otherwise works just like the other one.

**Comment by ErikSchierboom** [comment]
> All this PR does is add files. The current test runner still works, I just checked to make sure.

Great. Thanks for checking.

> I think I've got it much closer! I created an alternate run script called run-bb.sh which calls the babashka script instead of lein test and otherwise works just like the other one.

That output looks good! The only thing missing is the `test_code` field for tests (see https://github.com/exercism/docs/blob/main/building/tooling/test-runners/interface.md#test-code), but it's totally fine to do that in a later PR.

**Comment by ErikSchierboom** [comment]
@porkostomus You could perhaps try to update the tests in the `tests` folder, replace the `bin/run.sh` with your version and then running `bin/run-tests-in-docker.sh` to see if the tests pass.

**Comment by porkostomus** [comment]
Got the Docker stuff working, but just a couple of loose ends:

1. The test runner works by the bash script copying the babashka script into the project root so it can be executed there. But I'm not sure if this is ideal, I just happened to get it working that way.
2. The [test runner] tests are outputting diffs because I neglected to sort the JSON map. I suppose that order matters?

**Comment by ErikSchierboom** [comment]
> The test runner works by the bash script copying the babashka script into the project root so it can be executed there. But I'm not sure if this is ideal, I just happened to get it working that way.

This is fine. I don't think there's any performance hit so 👍 

> The [test runner] tests are outputting diffs because I neglected to sort the JSON map. I suppose that order matters?

Yes, the tests order matters in the JSON, as the test results are displayed on the website in the order in which they appear in the JSON. Ideally, the tests are ordered the way the tasks are ordered, so the task 1 tests first, then the task 2 tests, etc.

**Comment by porkostomus** [comment]
I needed to create a workaround for the tests being reported out of order. It turns out this behavior is baked deep into `clojure.test` so I found an efficient way to parse the test file and return a list of the tests in order, which we can then process into the `tests` array.

**Comment by porkostomus** [comment]
Since we have parsed the test file to get the test order, I went ahead and also grabbed the `test_code` info and added it to the output.

I think we're good for the version 2 spec.

File: bin/run.sh [hunk 1]
```diff
@@ -31,24 +31,23 @@ mkdir -p "${output_dir}"
...
+# pushd "${input_dir}" > /dev/null
 
 # Run the tests for the provided implementation file and redirect stdout and
 # stderr to capture it
-test_output=$(lein test 2>&1)
+test_output=$(./test-runner.clj "${slug}" "${input_dir}/" "${output_dir}" 2>&1)
 exit_code=$?
 
-popd > /dev/null
+# popd > /dev/null
```
Comment (ErikSchierboom) [inline bin/run.sh:41]: Maybe remove this commented line of code?

```suggestion
```

File: bin/run.sh [hunk 2]
```diff
@@ -31,24 +31,23 @@ mkdir -p "${output_dir}"
 
 echo "${slug}: testing..."
 
-pushd "${input_dir}" > /dev/null
+# pushd "${input_dir}" > /dev/null
```
Comment (ErikSchierboom) [inline bin/run.sh:34]: Maybe remove this commented line of code?

```suggestion
```

File: bin/run-tests.sh [hunk 3]
```diff
@@ -24,12 +24,12 @@ for test_dir in tests/*; do
...
-    bin/run.sh "${test_dir_name}" "${test_dir_path}" "${test_dir_path}"
+    bin/run.sh leap "${test_dir_path}" "${test_dir_path}"
 
     # Normalize the results file
-    sed -i -E \
-      -e "s~${test_dir_path}~/solution~g" \
-      "${results_file_path}"
+    #sed -i -E \
+    #  -e "s~${test_dir_path}~/solution~g" \
+    #  "${results_file_path}"
```
Comment (ErikSchierboom) [inline bin/run-tests.sh:32]: Maybe remove these commented lines of code?

```suggestion
```

**Review by ErikSchierboom — COMMENTED** [review]
Looking good! The only thing that I think is missing is the `message` for failing tests: https://github.com/exercism/docs/blob/main/building/tooling/test-runners/interface.md#message-1 The documentation did not make that clear (I've just submitted [this PR](https://github.com/exercism/docs/pull/232) to help clarify).

The `message` key should contain the message that the language's test runner outputs when it is run. Something like: `Assert failed. Expected "abc" but got "xyz"`

**Comment by porkostomus** [comment]
> The `message` key should contain the message that the language's test runner outputs when it is run. Something like: `Assert failed. Expected "abc" but got "xyz"`

I thought that looked familiar, the info was collected but never used. Added and populated the `message` key.

**Review by SleeplessByte — APPROVED** [review]

**Comment by SleeplessByte** [comment]
@porkostomus sad. No powers here. Let me ping Jeremy.

(cc https://github.com/exercism/clojure/pull/400#issuecomment-914641909)

**Review by ErikSchierboom — APPROVED** [review]

**Comment by iHiD** [comment]
Brilliant work on this :)

**Pull request merged**